lib/wasi/src/syscalls/mod.rs

```diff
@@ -1321,6 +1321,10 @@ pub fn fd_readdir<M: MemorySize>(
                             (entry.name.to_string(), stat.st_filetype, stat.st_ino)
                         }),
                 );
+                // adding . and .. special folders
+                // TODO: inode
+                entry_vec.push((".".to_string(), __WASI_FILETYPE_DIRECTORY, 0));
+                entry_vec.push(("..".to_string(), __WASI_FILETYPE_DIRECTORY, 0));
```

Comment:
Ummm... what about the root? We shouldn't add `..` to readdir in the root.
Can you also add a test case for this? Thanks

Reply:
But, isn't Root the next `match` here?

Reply:
Also, side comment. Do you have insight on why our fix is quite different from wasmtime's one?

https://github.com/bytecodealliance/wasmtime/pull/2494/files#diff-df2ef1370349b8385454adfabfe871fc51fd3c12debc7b0691b98ca19c5724deR309-R335

Reply:
What I hesitated is adding "." to a root readdir

Reply:
the root cause was completly different: it's not a buffer size issue here, it's the missing "." and "..". All the 201 folder appears correctly, just missing the special folders

Reply:
So no, adding "." to root is a bad idea, it breaks things.